docs/service_architecture.md

```diff
@@ -0,0 +1,168 @@
+
+## Example client-server-signer interaction
+
+The following diagram illustrates the interactions between the Notary client,
+sever, and signer:
+
+![Notary Service Sequence Diagram](images/metadata-sequence.svg)
+
+1. Notary server optionally supports authentication from clients using
+   [JWT](http://jwt.io/) tokens. This requires an authorization server that
+   manages access controls, and a cert bundle from this authorization server
+   containing the public key it uses to sign tokens.
+
+   If token authentication is enabled on Notary server, then any connecting
+   client that does not have a token will be redirected to the authorization
+   server.
+
+   Please see the docs for [Docker Registry v2 authentication](
+   https://github.com/docker/distribution/blob/master/docs/spec/auth/token.md)
+   for more information.
+
+2. The client will log in to the authorization server via basic auth over HTTPS,
+   obtain a bearer token, and then present the token to Notary server on future
+   requests.
+
+3. When clients uploads new metadata files, Notary server checks them against
+   any previous versions for conflicts, and verifies the signatures, checksums,
+   and validity of the uploaded metadata.
+
+4. Once all the uploaded metadata has been validated, Notary server
+   generates the timestamp (and maybe snapshot) metadata. It sends this
+   generated metadata to the Notary signer to be signed.
+
+5. Notary signer gets the requisite encrypted private keys from its database if
+   it has them, decrypts the keys, and uses them to sign the metadata. If
+   successful, it sends the signatures back to Notary server.
+
+6. Notary server stores all the signed metadata, both client-uploaded and
+   server-generated metadata, in the TUF database. The generated timestamp and
+   snapshot metadata certify that the metadata files the client uploaded are the
+   most recent for that repository.
+
+   Finally, Notary server will notify the client that their upload was successful.
+
+7. The client can now immediately download the latest metadata from the server,
+   using the still-valid bearer token to connect. Notary server only needs to
+   obtain the metadata from the database, since none of the metadata is expired
+   yet.
+
+   If the timestamp were expired, for example, Notary server would go through
+   the entire sequence where it generates a new timestamp, request Notary signer
+   for a signature, stores the newly signed timestamp in the database. It then
+   sends this new timestamp, along with the rest of the stored metadata, to the
+   requesting client.
+
+
+## Threat model
+
+Both the server and the signer are possible attack vectors. This section
+discusses how the architecture responds when compromised.
+
+## Notary server compromise
+
+In the event of a Notary server compromise, an attacker would have access to
+the metadata stored in the database as well as access to Notary signer to
```

Comment:
as well as access to the credentials to communicate with Notary signer, and therefore, access to arbitrary signing operations with any key the Signer holds.